scripts/build_climate_projections.py

```diff
@@ -0,0 +1,480 @@
+# -*- coding: utf-8 -*-
+# SPDX-FileCopyrightText:  PyPSA-Earth and PyPSA-Eur Authors
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+# -*- coding: utf-8 -*-
+"""
+Creates a dataset corresponding to the projected climate in a requested year.
+
+Relevant Settings
+-----------------
+
+.. code:: yaml
+
+    projection:
+        climate_scenario:
+        present_year:
+        future_year:
+        years_window:
+        gcm_selection:
+
+Inputs
+------
+
+- ``cutouts/cutout.nc``: confer :ref:`cutout`, a cutout file produced by altile
+- ``data/cmip6_avr.nc``: confer :ref:`cmip6`, a global-scale dataset of the global climate
+CMIP6 projections aggregated into IPCC Global Climate Atlas
+
+Outputs
+-------
+
+- ``cutouts/proj-{cutout}.nc"``: confer :ref:`proj`, a cutout modified to account for future
+climate conditions
+
+Description
+-----------
+
+The rule :mod:`build_climate_projections` creates a cutout file which corresponds to
+a requested year in the future. The parameter projections is calculated combining data
+for cutout.nc which is assumed to be representative data sample for the past climate and
+an ensemble of CMIP6 global climate models to account for the future climate.
+
+A morphing approach is implemented meaning that the projection value x_future is calculated
+data for a representative period present_year in the past x_past using a combination
+of shift dx and stretch a:
+
+x_future = x_past + dx + a(x_past - <x_past>|month),
+
+where
+x_future and x_past have the highest temporal resolution,
+<x_past>|month is the monthly average for a considered month.
+
+The procedure is applied for each month to account for seasonality. Snapshots settings
+are used to define for which months to be used in calculating a climate projection.
+
+Methodology notes
+-----------------
+Originally the morphing approach originates from buildings physics, and is still in active
+use for detailed simulation of heat transfer in buildings. See, for example:
+
+J.B. Dias, G.C. da Graça, P.M.M. Soares (2020) Comparison of methodologies for
+generation of future  building thermal energy simulation. Energy and Buildings,
+vol. 206, 109556, https://doi.org/10.1016/j.enbuild.2019.109556.
+
+J. Pulkkinen, J.-N. Louis (2021) Near- and medium-term hourly morphed mean and extreme future temperature datasets for Jyväskylä, Finland, for building thermal energy demand simulations. Data in Brief, vol. 37, 107209, https://doi.org/10.1016/j.dib.2021.107209.
+"""
+
+
+import datetime as dt
+import os
+
+import atlite
+import geopandas as gpd
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import pypsa
+import xarray as xr
+from _helpers import configure_logging, create_logger
+from shapely.geometry import LineString as Line
+from shapely.geometry import Point
+
+logger = create_logger(__name__)
+
+
+def crop_cmip6(cmip6_xr, cutout_xr):
```

Comment:
The docstrings generally miss the description of the output. could be nice to have, what do you think?

Reply:
Absolutely! Fixed :)